pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/performance/AvoidNonRestrictiveQueriesRule.java

```diff
@@ -0,0 +1,45 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.rule.performance;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import net.sourceforge.pmd.lang.apex.ast.ASTSoqlExpression;
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexRule;
+
+public class AvoidNonRestrictiveQueriesRule extends AbstractApexRule {
+    private static final Pattern RESTRICTIVE_PATTERN = Pattern.compile("(where )|(limit )", Pattern.CASE_INSENSITIVE);
+    private static final Pattern SELECT_PATTERN = Pattern.compile("(select )", Pattern.CASE_INSENSITIVE);
+    
+    public AvoidNonRestrictiveQueriesRule() {
+        setProperty(CODECLIMATE_CATEGORIES, "Performance");
+        setProperty(CODECLIMATE_REMEDIATION_MULTIPLIER, 100);
+        setProperty(CODECLIMATE_BLOCK_HIGHLIGHTING, false);
+    }
+
+    @Override
+    public Object visit(ASTSoqlExpression node, Object data) {
+        Integer occurencesSelect = 0;
+        Integer occurencesWhereOrLimit = 0;
+        
+        Object o = node.getNode().getRawQuery();
+        if (o instanceof String) {
+            String query = (String) o;
+            Matcher matcherRestrictive = RESTRICTIVE_PATTERN.matcher(query);
+            Matcher matcherSelect = SELECT_PATTERN.matcher(query);
+            while (matcherRestrictive.find()) {
+                occurencesWhereOrLimit++;
+            }
+            while (matcherSelect.find()) {
+                occurencesSelect++;
+            }
+            if (occurencesSelect > occurencesWhereOrLimit) {
```

Comment:
if I'm getting this right, this approach would consider no violations when comming across a query such as:

```
[ select id, (SELECT Id FROM Contact where id = 1 limit 1) from account ];
```

But we are effectively retrieving the whole account table...

Similarly, the query

```
[ select id, (SELECT Id FROM Contact) from account limit 1 ];
```

would be flagged. Even though the query may not be the most efficient for the db, as for governor limits, we are always retrieving 1 row, which should be ok (according to my research, the limit is 2000 per SOSL query).


I fear we will need to better parse the query to accurately detect violations for this rule.

Reply:
IMO, the rule of thumb is to ignore sub-queries, and to evaluate only against the top-level query. So the first query is bad and the second is good. (Meaning that the flagging actions in this PR are backwards.)